05.ls/ls04.rb

```diff
@@ -0,0 +1,214 @@
+# frozen_string_literal: true
+
+require 'etc'
+require 'optparse'
+
+SIX_MONTH = 15_552_000
+ROW_NUM = 3
+ROW_MAX_WIDTH = 24
+
+def main
+  paths = distingish_options
+  total_block = acquire_blocks(paths)
+  file_modes = acquire_file_modes
+  links = acquire_links(paths)
+  names = acquire_names(paths)
+  groups = acquire_groups(paths)
+  sizes = acquire_sizes(paths)
+  months = acquire_months(paths)
+  days = acquire_days(paths)
+  times = acquire_times(paths)
+  files = acquire_files
+
+  puts "total #{total_block}"
+  file_modes.zip(links, names, groups, sizes, months, days, times, files).each do |row|
+    puts row.join
+  end
+end
+
+def distingish_options
+  opt = OptionParser.new
+  opt.on('-l')
+  opt.parse(ARGV)
+  if ARGV == ['-l']
+    acquire_absolute_paths
+  else
+    all_files = Dir.glob('*').sort
+    files_in_columns = get_transposed_all_files(all_files)
+    display(files_in_columns)
+    exit
+  end
+end
+
+# オプションなし
+
+def get_transposed_all_files(all_files)
+  all_files.push(' ') while all_files.length % ROW_NUM != 0
+  column_num = all_files.length / ROW_NUM
+  transposed_files = all_files.each_slice(column_num).to_a.transpose
+  transposed_files.first(column_num).each do |column|
+    ROW_NUM.times do |index|
+      column[index] += ' ' * (ROW_MAX_WIDTH - column[index].length)
+    end
+  end
+end
+
+def display(files_in_columns)
+  files_in_columns.each do |column|
+    puts column.join
+  end
+end
+
+#-lオプションありの場合はここから
+
+# 絶対パスの取得
+def acquire_absolute_paths
+  files = Dir.glob('*').sort
+  current_path = Dir.getwd
+  files.map do |file|
+    "#{current_path}/#{file}"
+  end
+end
+
+# 総ブロック数の取得
+def acquire_blocks(paths)
+  blocks = paths.map do |path|
+    File.stat(path).blocks
+  end
+  blocks.sum
+end
+
+# ファイルモードの取得
+def acquire_file_modes
+  paths = acquire_absolute_paths
+  nums = acquire_file_mode_nums(paths)
+  adjusted_nums = adjust_file_mode_nums_length(nums)
+  convert_to_letter(adjusted_nums)
+end
+
+def acquire_file_mode_nums(paths)
+  paths.map do |path|
+    format('0%o', File.stat(path).mode)
+  end
+end
+
+def adjust_file_mode_nums_length(nums)
+  nums.map do |num|
+    if num.length < 7
+      num.insert(0, '0')
+    else
+      num
+    end
+  end
+end
+
+def convert_to_letter(adjusted_nums)
+  file_types_nums = { '001' => 'p', '002' => 'c', '004' => 'd', '006' => 'b', '010' => '-', '012' => 'l', '014' => 's' }
```

Comment:
こういう風に書くと読みやすいのでおすすめです！

```suggestion
  file_types_nums = {
    '001' => 'p',
    '002' => 'c',
    '004' => 'd',
    '006' => 'b',
    '010' => '-',
    '012' => 'l',
    '014' => 's'
  }
```